device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/https/HttpsBatchMessage.java

```diff
@@ -41,42 +43,37 @@
     /** The current number of messages in the batch. */
     private int numMsgs;
 
-    /** Constructor. Initializes the batch body as an empty JSON array. */
-    public HttpsBatchMessage()
+    public HttpsBatchMessage(List<HttpsSingleMessage> messageList) throws IotHubSizeExceededException
     {
-        // Codes_SRS_HTTPSBATCHMESSAGE_11_001: [The constructor shall initialize the batch message with the body as an empty JSON array.]
-        this.batchBody = "[]";
-        this.numMsgs = 0;
-    }
+        StringBuilder batchBodyBuilder = new StringBuilder();
+        batchBodyBuilder.append('[');
 
-    /**
-     * Adds a message to the batch.
-     *
-     * @param msg the message to be added.
-     *
-     * @throws IotHubSizeExceededException if adding the message causes the
-     * batched message to exceed 256 kb in size. The batched message will remain
-     * as if the message was never added.
-     */
-    public void addMessage(HttpsSingleMessage msg) throws IotHubSizeExceededException
-    {
-        String jsonMsg = msgToJson(msg);
-        // Codes_SRS_HTTPSBATCHMESSAGE_11_002: [The function shall add the message as a JSON object appended to the current JSON array.]
-        String newBatchBody = addJsonObjToArray(jsonMsg, this.batchBody);
+        boolean atLeastOneMessage = false;
+        for (HttpsSingleMessage message : messageList)
+        {
+            addJsonToStringBuilder(message, batchBodyBuilder);
+            batchBodyBuilder.append(','); // comma to separate each object in the json array
+            this.numMsgs++;
+            atLeastOneMessage = true;
+        }
+
+        if (atLeastOneMessage)
+        {
+            batchBodyBuilder.deleteCharAt(batchBodyBuilder.length() - 1); // remove the final comma from the list
```

Comment:
Another approach to consider, that doesn't requiring removing a character, but still has the overhead of one boolean checked on each iteration in the loop above (but only requires setting once), is to conditionally add the comma only after the first time.

So something like:
```Java
boolean isSubsequentMessage = false;
        for (HttpsSingleMessage message : messageList)
        {
            if (isSubsequentMessage) {
                batchBodyBuilder.append(','); // comma to separate each object in the json array
            } else {
                isSubsequentMessage = true;
            }
            addJsonToStringBuilder(message, batchBodyBuilder);
            this.numMsgs++;
        }
```

Reply:
Another thought here too is to consider [String.join](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#join-java.lang.CharSequence-java.lang.Iterable-).

In this instance you'd move the JSON string creation to a method in the `HttpsSingleMessage` class, create the List<String> in `sendMessage` and create a `List<String>` constructor to be used. 

A pro of this is we are using something built into the SDK and it makes it more readable; plus less lines of code. As well you're creating a method to be reused anytime you might want to spit out the JSON of a HttpsSingleMessage.

Some cons are you're moving the string creation up a level into `sendMessage` which may not want and you'd be adding a new method to the  `HttpsSingleMessage` class and it may make the batch message creation too disjointed.